### PR TITLE
soc: arc: remove <soc.h> usage

### PR DESCRIPTION
**Pull request by gmarull**
<soc.h> has traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases, it includes HALs,
in some others, it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL (bypassing the purpose of DT in some cases…).
To make things worse, it is being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

NOTE: I have zero ARC knowledge, so some patches may not contain the best solution (specially the DT part).

Related to https://github.com/zephyrproject-rtos/zephyr/pull/46147 and https://github.com/zephyrproject-rtos/zephyr/issues/41543

### CODE REVIEW

**Review by andyross — DISMISSED** [review]

File: soc/arc/snps_arc_iot/soc_irq.h
Comment (ruuddw) [inline soc/arc/snps_arc_iot/soc_irq.h:11]: This file had as a convenience all interrupts defined. In Zephyr, not all devices/interrupts are supported, so it makes sense only the interrupts used where added to the dts files.
That will have to be extended if more devices are added/supported in the future.

File: soc/arc/snps_nsim/soc.c
Comment (ruuddw) [inline soc/arc/snps_nsim/soc.c:19]: This doesn't look right, the IDU (interrupt dispatch unit) is only available/supported for SMP configurations.
Why was the ifdef removed?

File: tests/kernel/context/src/main.c
Comment (ruuddw) [inline tests/kernel/context/src/main.c:31]: so for posix soc.h is still allowed, but for ARC not ;-)

**Review by ruuddw — CHANGES_REQUESTED** [review]
Thanks for the removal of many soc.h that where not really needed.
The interrupts number definitions that where moved to dts are not configurable and will never change (ARC architecture defined). But, for consistency it makes sense to move them to DTS.
Just one change that I think is not correct, see comments (SMP ifdef that was removed).

I don't have all the boards currently available, so couldn't test all - but, we'll cover that during this release cycle and fix if we find something later.

File: soc/arc/snps_arc_iot/soc_irq.h
Comment (gmarull) [reply soc/arc/snps_arc_iot/soc_irq.h:11]: if more devices are added, interrupts should be defined in Devicetree (together with their priority), not in C files anymore. Zephyr coding guidelines state that there should not be any dead code, ref. https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html. Many ARM SoC HALs also define all interrupt numbers, but in Zephyr they're not used (and if they do it should be fixed).

File: soc/arc/snps_nsim/soc.c
Comment (gmarull) [reply soc/arc/snps_nsim/soc.c:19]: The code has been changed so that this file is only compiled if `CONFIG_SMP=y`. As it was before, the file was always included in the build but had no code when `CONFIG_SMP=n`. See https://github.com/zephyrproject-rtos/zephyr/pull/46585/commits/cf250a2d017e060943121badab8d439d25909576

File: tests/kernel/context/src/main.c
Comment (gmarull) [reply tests/kernel/context/src/main.c:31]: It's not about ARC, please read the PR for more context. ARM64 has already been cleanup, and others will gradually come. I've started with small platforms, others like ARM will take more time.

Comment (andyross) [reply tests/kernel/context/src/main.c:31]: Drive-by comment:  this test is just awful, literally having to punch a hole through into every timer driver in the whole tree, which is why the soc.h madness exists here in a platform-dependent way.

I had a patch sitting around as part of the MTL audio DSP work that just defined a global variable to hold the IRQ (which would then be set by the drivers, ideally in a static initializer such that it optimizes out if unused), which eliminated all the complexity.  But I never got around to making it work for all the drivers.

File: soc/arc/snps_nsim/soc.c
Comment (nashif) [reply soc/arc/snps_nsim/soc.c:19]: that still does not look right IMO, we probably should leave it as is. Why would generic soc.c be dependent on CONFIG_SMP, soc.c is there for many things, not only enabling SMP. I can see this apply to many other files enable a specific feature, but here I would consider this over-optimization and a trap if someone adds something to soc.c for reasons  other than SMP.

**Review by nashif — CHANGES_REQUESTED** [review]

File: soc/arc/snps_nsim/soc.c
Comment (gmarull) [reply soc/arc/snps_nsim/soc.c:19]: > that still does not look right IMO, we probably should leave it as is. Why would generic soc.c be dependent on CONFIG_SMP, soc.c is there for many things, not only enabling SMP. I can see this apply to many other files enable a specific feature, but here I would consider this over-optimization and a trap if someone adds something to soc.c for reasons  other than SMP.

Some socs have more than one C file, here soc.c is not a generic source, so it can likely be renamed to smp.c to avoid confusions. C files becoming empty under certain conditions and still being fed to the compiler isn't a great practice, it just increases the number of files to parse for nothing. cc: @tejlmand

File: tests/kernel/context/src/main.c
Comment (gmarull) [reply tests/kernel/context/src/main.c:31]: > Drive-by comment:  this test is just awful, literally having to punch a hole through into every timer driver in the whole tree, which is why the soc.h madness exists here in a platform-dependent way.
> 
> 
> 
> I had a patch sitting around as part of the MTL audio DSP work that just defined a global variable to hold the IRQ (which would then be set by the drivers, ideally in a static initializer such that it optimizes out if unused), which eliminated all the complexity.  But I never got around to making it work for all the drivers.

It would definitely be nice to clean this up.

File: soc/arc/snps_nsim/soc.c
Comment (evgeniy-paltsev) [reply soc/arc/snps_nsim/soc.c:19]: Indeed, let's keep this as it is.

Comment (ruuddw) [reply soc/arc/snps_nsim/soc.c:19]: Agreed, I missed the condition was moved to the CMakeList.txt.

**Review by ruuddw — DISMISSED** [review]

**Review by andyross — DISMISSED** [review]
Refresh +1

File: soc/arc/snps_nsim/soc.c
Comment (nashif) [reply soc/arc/snps_nsim/soc.c:19]: still not convinced and, the code in the file is what soc.c is for, it happens to be all SMP specific, fine, does not mean we need to change the name of the file, especially for this type of simulation platforms where code come and goes and soc.c is for this exact puropse.

if we are looking for "performance" gains, I would be looking elsewhere leaving this the way it is.

Comment (ruuddw) [reply soc/arc/snps_nsim/soc.c:19]: I don't have a strong opinion for one or the other, is there a convention used in other SoCs that we can align with? Consistency over different SoCs would be least confusing and minimizing future problems/mistakes on code updates/additions. For this specific 'SoC' I don't expect we'll add many SoC specific initializations in the future, as this is a CPU-focused simulator.

Comment (gmarull) [reply soc/arc/snps_nsim/soc.c:19]: There's probably a similar example on certain ARM SoCs that configure the MMU regions. They have a `mmu_regions.c` that is included only if `CONFIG_ARM_MMU=y`. I think the pattern here is similar. You can actually grep for `zephyr_sources_ifdef` in `soc/`, I see many other examples like for PM. But `soc/` is likely not the best place to look for consistency (see <soc.h>).

Can please maintainers (platform, build system) decide then? I would rather not block progress because of personal preferences here.

File: soc/arc/snps_nsim/smp.c
Comment (ruuddw) [inline soc/arc/snps_nsim/smp.c:18]: if this indeed moves to a module called smp.c, then let's not call this arc_nsim_init (which made sense for soc.c), but something smp specific.

File: soc/arc/snps_nsim/soc.c
Comment (ruuddw) [reply soc/arc/snps_nsim/soc.c:19]: Well, the main contribution of this PR is the soc.h cleanup. If you keep the soc.c SMP stuff like it was before, I think the PR will be quickly approved. If you feel strong about the SMP init part, you can create an issue for it and/or a separate PR.

**Comment by evgeniy-paltsev** [comment]
FYI: I've run in-house verification on this, waiting for results.

File: dts/arc/synopsys/emsk.dtsi
Comment (evgeniy-paltsev) [inline dts/arc/synopsys/emsk.dtsi:46]: Hmm that looks not entirely correct - this interrupt exists not on all boards, see (from soc.h)

```
#if defined(CONFIG_BOARD_EM_STARTERKIT_R23) && defined(CONFIG_SOC_EMSK_EM7D)
#define IRQ_SEC_TIMER0				20
#endif /* CONFIG_BOARD_EM_STARTERKIT_R23 && CONFIG_SOC_EMSK_EM7D */
```

but dts/arc/synopsys/emsk.dtsi is used for all boards.

And on some configurations this irq line is for `IRQ_CORE_DMA_COMPLETE` interrupt.

File: soc/arc/snps_nsim/smp.c
Comment (nashif) [reply soc/arc/snps_nsim/smp.c:18]: exactly! That is why my preference is not to change anything and just leave it as soc.c, this sys_init stuff really belongs into soc.c and moving things around like this for the purpose of cleaning up (the PR title still says "remove soc.h usage") and this change has nothing to do with soc.h usage and now we are getting into renaming stuff etc etc....

Oh well...

**Comment by evgeniy-paltsev** [comment]
Hmm, got huge amount of runtime failures on all platforms except qemu. This require additional investigation.

File: dts/arc/synopsys/emsk.dtsi
Comment (gmarull) [reply dts/arc/synopsys/emsk.dtsi:46]: `IRQ_SEC_TIMER0` seems to be used only in `drivers/timer/arcv2_timer0.c`, only if `CONFIG_ARC_SECURE_FIRMWARE=y`, so from what I understand, certain boards can't enable this option. This needs to be specified somehow (probably not by adding board dependencies, but an auxiliary Kconfig option like `HAS_ARC_SEC_TIMER` or whatever else, selected at SoC level).

I can still define this node at board level, making it 100% equivalent to the previous code, but to me, this shows that there's still an issue somewhere.

Comment (gmarull) [reply dts/arc/synopsys/emsk.dtsi]: see latest change: [`ff72f58` (#46585)](https://github.com/zephyrproject-rtos/zephyr/pull/46585/commits/ff72f58de4a019784c6629f5173dfd64f5a4ad2c)

File: soc/arc/snps_nsim/smp.c
Comment (gmarull) [reply soc/arc/snps_nsim/smp.c:18]: > exactly! That is why my preference is not to change anything and just leave it as soc.c, this sys_init stuff really belongs into soc.c and moving things around like this for the purpose of cleaning up (the PR title still says "remove soc.h usage") and this change has nothing to do with soc.h usage and now we are getting into renaming stuff etc etc....
> 
> Oh well...

I'm not sure how such reviewing behavior helps to make progress. It's nice to talk about consistency, but it's also nice to help to make progress on PRs that precisely help towards that goal. Can we please focus on the technical aspects and comment on latest changes? I see no comments related to [`31e1a9f` (#46585)](https://github.com/zephyrproject-rtos/zephyr/pull/46585/commits/31e1a9f2a47c7afb037b70fd6df08f0751193c2b)

Comment (ruuddw) [reply soc/arc/snps_nsim/smp.c:18]: My comment to not only rename the filename but also the function name (arc_nsim_init) to make it SMP specific was about this commit. There are many improvements to get rid of soc.h in this PR, but avoiding the sometimes empty soc.c meets different opinions. If you prefer to include this commit, please rename arc_nsim_init, e.g. to arc_nsim_smp_init.

Comment (gmarull) [reply soc/arc/snps_nsim/smp.c]: > My comment to not only rename the filename but also the function name (arc_nsim_init) to make it SMP specific was about this commit. There are many improvements to get rid of soc.h in this PR, but avoiding the sometimes empty soc.c meets different opinions. If you prefer to include this commit, please rename arc_nsim_init, e.g. to arc_nsim_smp_init.

done

Comment (ruuddw) [reply soc/arc/snps_nsim/smp.c]: Thanks, I'm ok with this.

**Comment by ruuddw** [comment]
> Hmm, got huge amount of runtime failures on all platforms except qemu. This require additional investigation.

@evgeniy-paltsev Do the problems still occur with latest updates? Any idea about the cause?

File: boards/arc/nsim/nsim.dtsi
Comment (ruuddw) [inline boards/arc/nsim/nsim.dtsi:36]: typo, I think the cause for the issues @evgeniy-paltsev mentioned. Should be timer1: timer1.

**Review by ruuddw — CHANGES_REQUESTED** [review]

**Review by ruuddw — DISMISSED** [review]

**Comment by evgeniy-paltsev** [comment]
Hi @gmarull ,
the in-house verification passes now. In general this looks good. The only thing that bothers me is that we adding
device tree entry with devices which doesn't exist everywhere (`intercore-interrupt-unit`, `sectimer0`) to generic device tree include files - previously we've define this interrupts line conditionally.

Given that clean implementation require significant device tree rework and given that it shouldn't break anything now (as we only get interrupt line from this devices) I'm OK with keeping proposed implementation.

So, may I ask you to add this comment to the places where you add `intercore-interrupt-unit` or `sectimer0` device tree entries?

```
/* FIXME: this device may not exist on some platforms which includes this dtsi! */
```

Thanks!

**Comment by gmarull** [comment]
@evgeniy-paltsev @ruuddw please check latest changes

**Review by ruuddw — APPROVED** [review]
Thanks!
(but let's wait for Evgeniy's final testing/ack)

**Comment by gmarull** [comment]
@evgeniy-paltsev gentle ping

**Comment by gmarull** [comment]
> Hi @gmarull,
> 
> I've tested this finally - everything is OK now, thanks!

can you approve then?

**Review by evgeniy-paltsev — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**
